## This data science project aims to predict possibility of road accident and its severity. This will help road users to decide on a journey and driving habit.  The dataset contains various details like severity, collision type, injuries, fatality, location, etc. 

In [2]:
import pandas as pd
import numpy as np

### About dataset

This dataset is about past road accident and severity. The __data_collisions.csv__ data set includes details of 194672 accident records, location of accident and severity. It includes some of the following fields:

| Field          | Description                                                                           |
|----------------|---------------------------------------------------------------------------------------|
| Location       | location of collision                                                                 |
| Severity       | Code the correspond to the severity of the collision                                  |
| severity Desc  | A detailed description of the severity of the collision                               | 
| Ped count      | The number of pedestrians                                                             |
| SeriousInjuries| The number of serious injuries in the collision                                       |
| Injuries       | The number of total injuries in the collision                                         |
| Fatalities     | The number of fatalities in the collision                                             |
| Weather        | Wseather condition during the time of the collision                                   |
| COLLISIONTYPE  | collision type                                                                        |
| PERSONCOUNT    |The total number of people involved in the collision                                   |
| PEDCYLCOUNT    | The number of bicycles involved in the collision. This is entered by the state.       | 
| VEHCOUNT       | The number of vehicles involved in the collision.This is entered by the state.        |
| INCDATE        | The date of the incident                                                              |
| INCDTTM        | The date and time of the incident.                                                    |
| JUNCTIONTYPE   | Category of junction at which collision took place                                    |
| SDOT_COLCODE   | A code given to the collision by SDOT.                                                |
| SDOT_COLDESC   | A description of the collision corresponding to the collision code.                   |
| INATTENTIONIND | Whether or not collision was due to inattention. (Y/N                                 |
| UNDERINFL      | Whether or not a driver involved was under the influence of drugs or alcohol.         | 
| ROADCOND       | The condition of the road during the collision.                                       |
| LIGHTCOND      | The light conditions during the collision.                                            |
| PEDROWNOTGRNT  | Whether or not the pedestrian right of way was not granted. (Y/N)                     |
| SPEEDING       | Whether or not speeding was a factor in the collision. (Y/N)                          |
| CROSSWALKKEY   | A key for the crosswalk at which the collision occurred.                              |

### Load data from CSV File

In [6]:
df=pd.read_csv("Data-Collisions.csv", parse_dates=["INCDATE"])
df.info()
df.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194673 entries, 0 to 194672
Data columns (total 38 columns):
SEVERITYCODE      194673 non-null int64
X                 189339 non-null float64
Y                 189339 non-null float64
OBJECTID          194673 non-null int64
INCKEY            194673 non-null int64
COLDETKEY         194673 non-null int64
REPORTNO          194673 non-null object
STATUS            194673 non-null object
ADDRTYPE          192747 non-null object
INTKEY            65070 non-null float64
LOCATION          191996 non-null object
EXCEPTRSNCODE     84811 non-null object
EXCEPTRSNDESC     5638 non-null object
SEVERITYCODE.1    194673 non-null int64
SEVERITYDESC      194673 non-null object
COLLISIONTYPE     189769 non-null object
PERSONCOUNT       194673 non-null int64
PEDCOUNT          194673 non-null int64
PEDCYLCOUNT       194673 non-null int64
VEHCOUNT          194673 non-null int64
INCDATE           194673 non-null datetime64[ns, UTC]
INCDTTM           194673 

SEVERITYCODE           X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO  \
0             2 -122.323148  47.703140         1    1307       1307  3502005   
1             1 -122.347294  47.647172         2   52200      52200  2607959   
2             1 -122.334540  47.607871         3   26700      26700  1482393   
3             1 -122.334803  47.604803         4    1144       1144  3503937   
4             2 -122.306426  47.545739         5   17700      17700  1807429   

    STATUS      ADDRTYPE   INTKEY  ... ROADCOND                LIGHTCOND  \
0  Matched  Intersection  37475.0  ...      Wet                 Daylight   
1  Matched         Block      NaN  ...      Wet  Dark - Street Lights On   
2  Matched         Block      NaN  ...      Dry                 Daylight   
3  Matched         Block      NaN  ...      Dry                 Daylight   
4  Matched  Intersection  34387.0  ...      Wet                 Daylight   

  PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0           NaN         NaN      NaN         10   
1           NaN   6354039.0      NaN         11   
2           NaN   4323031.0      NaN         32   
3           NaN         NaN      NaN         23   
4           NaN   4028032.0      NaN         10   

                                          ST_COLDESC  SEGLANEKEY  \
0                                  Entering at angle           0   
1  From same direction - both going straight - bo...           0   
2                             One parked--one moving           0   
3                   From same direction - all others           0   
4                                  Entering at angle           0   

   CROSSWALKKEY  HITPARKEDCAR  
0             0             N  
1             0             N  
2             0             N  
3             0             N  
4             0             N  

[5 rows x 38 columns]

In [8]:
df.describe(include="all")

SEVERITYCODE              X              Y       OBJECTID  \
count   194673.000000  189339.000000  189339.000000  194673.000000   
unique            NaN            NaN            NaN            NaN   
top               NaN            NaN            NaN            NaN   
freq              NaN            NaN            NaN            NaN   
first             NaN            NaN            NaN            NaN   
last              NaN            NaN            NaN            NaN   
mean         1.298901    -122.330518      47.619543  108479.364930   
std          0.457778       0.029976       0.056157   62649.722558   
min          1.000000    -122.419091      47.495573       1.000000   
25%          1.000000    -122.348673      47.575956   54267.000000   
50%          1.000000    -122.330224      47.615369  106912.000000   
75%          2.000000    -122.311937      47.663664  162272.000000   
max          2.000000    -122.238949      47.734142  219547.000000   

               INCKEY      COLDETKEY REPORTNO   STATUS ADDRTYPE  \
count   194673.000000  194673.000000   194673   194673   192747   
unique            NaN            NaN   194670        2        3   
top               NaN            NaN  1782439  Matched    Block   
freq              NaN            NaN        2   189786   126926   
first             NaN            NaN      NaN      NaN      NaN   
last              NaN            NaN      NaN      NaN      NaN   
mean    141091.456350  141298.811381      NaN      NaN      NaN   
std      86634.402737   86986.542110      NaN      NaN      NaN   
min       1001.000000    1001.000000      NaN      NaN      NaN   
25%      70383.000000   70383.000000      NaN      NaN      NaN   
50%     123363.000000  123363.000000      NaN      NaN      NaN   
75%     203319.000000  203459.000000      NaN      NaN      NaN   
max     331454.000000  332954.000000      NaN      NaN      NaN   

               INTKEY  ... ROADCOND LIGHTCOND PEDROWNOTGRNT    SDOTCOLNUM  \
count    65070.000000  ...   189661    189503          4667  1.149360e+05   
unique            NaN  ...        9         9             1           NaN   
top               NaN  ...      Dry  Daylight             Y           NaN   
freq              NaN  ...   124510    116137          4667           NaN   
first             NaN  ...      NaN       NaN           NaN           NaN   
last              NaN  ...      NaN       NaN           NaN           NaN   
mean     37558.450576  ...      NaN       NaN           NaN  7.972521e+06   
std      51745.990273  ...      NaN       NaN           NaN  2.553533e+06   
min      23807.000000  ...      NaN       NaN           NaN  1.007024e+06   
25%      28667.000000  ...      NaN       NaN           NaN  6.040015e+06   
50%      29973.000000  ...      NaN       NaN           NaN  8.023022e+06   
75%      33973.000000  ...      NaN       NaN           NaN  1.015501e+07   
max     757580.000000  ...      NaN       NaN           NaN  1.307202e+07   

       SPEEDING ST_COLCODE              ST_COLDESC     SEGLANEKEY  \
count      9333     194655                  189769  194673.000000   
unique        1        115                      62            NaN   
top           Y         32  One parked--one moving            NaN   
freq       9333      27612                   44421            NaN   
first       NaN        NaN                     NaN            NaN   
last        NaN        NaN                     NaN            NaN   
mean        NaN        NaN                     NaN     269.401114   
std         NaN        NaN                     NaN    3315.776055   
min         NaN        NaN                     NaN       0.000000   
25%         NaN        NaN                     NaN       0.000000   
50%         NaN        NaN                     NaN       0.000000   
75%         NaN        NaN                     NaN       0.000000   
max         NaN        NaN                     NaN  525241.000000   

        CROSSWALKKEY  HITPARKEDCAR  
count 

initial obeservation shows that the data is imbalance. the maximum non-null column count is 194673 while the non-null column is 4667

In [10]:
df['SEVERITYCODE'].value_counts()

1    136485
2     58188
Name: SEVERITYCODE, dtype: int64

In [12]:
df['ROADCOND'].value_counts()

Dry               124510
Wet                47474
Unknown            15078
Ice                 1209
Snow/Slush          1004
Other                132
Standing Water       115
Sand/Mud/Dirt         75
Oil                   64
Name: ROADCOND, dtype: int64

#### code and descriptions provided by the state that describes the collision.

In [15]:
df['ST_COLDESC'].value_counts()

One parked--one moving                                                                   44421
Entering at angle                                                                        34674
From same direction - both going straight - one stopped - rear-end                       25771
Fixed object                                                                             13554
From same direction - both going straight - both moving - sideswipe                      12777
From opposite direction - one left turn - one straight                                   10324
From same direction - both going straight - both moving - rear-end                        7629
Vehicle - Pedalcyclist                                                                    4701
From same direction - all others                                                          4537
From same direction - one left turn - one straight                                        3093
From same direction - one right turn - one straigh

In [18]:
df2=df[['SEVERITYCODE','ROADCOND']]

In [19]:
import matplotlib.pyplot as plt 
import seaborn as sns 
%matplotlib inline